### DISCUSSION TITLE
Solver not applicable (gurobi does not support general nonlinearities).

### DISCUSSION THREAD

**Discussion by BQKS** [yalmip/YALMIP#1401] [Optimization modelling] [answered]
Objective function contains Logarithmic and min function, constraint contains second-order cone, when using Yalmap+Gurobi+Matlab solves such problem, results display: Solver not applicable (gurobi does not support general nonlinearities). Are there some good methods to solve such problem when using Yalmap in Gurobi+Matlab?

**Comment by johanlofberg** [comment, marked as answer]
https://yalmip.github.io/command/interp1/#:~:text=Example%20(Linear%20optimization)

**Reply by BQKS** [reply]
Teacher, I'm going to use 'graph' to approximate the exponential function, but the mathematical model only need the low bound of the  hypograph at every abscissa point, and how to solve such problem?

**Reply by johanlofberg** [reply]
unclear what you are talking about. since you have exp and talk about graph model, it will be an epigraph representation. if you require the solution to be at a sampling point, it is not a convex problem as it simply is a selection (i.e requires binary variables)

**Reply by BQKS** [reply]
But there is the other question, when I use interp1 to express the exponential function, then bring into the mathematic model, the Gurobi still display (gurobi does not support general nonlinearities), whether I misunderstood your first comment（use interp1 to solve the nonlinearities）.

**Reply by johanlofberg** [reply]
Then you aren't using the flags 'lp', 'graph', 'milp' or 'sos2' in the interp1 operator

**Reply by BQKS** [reply]
Yes, I aren‘t，I use ’spline‘ in the interp1, but I don't understand using 'graph' to approximate the nonlinearities function. If the nonlinearities function less than or equal to another function, then putting the ’graph‘ Approximate function to replace the nonlinearities function into the Inequality constraint, but using 'graph' will produce greater than or equal to the nonlinearities function (assuming the nonlinearities function is convex function), bring in the Inequality constraint, whether produce the problem?

**Reply by johanlofberg** [reply]
Well you can't. If you use default options, then the operator will simply be the nonlinear function 'interp1' just like using 'sin' or which ever function, using that particular interpolation option, which is impossible when you use gurobi as gurobi does not support arbitrary nonlinear functions. When you use the 'lp'/'sos2'/'graph'/'milp' flags, you ask YALMIP to derive a (MI)LP model of the piecewise affine approximation

**Reply by BQKS** [reply]
Following is the use of 'sos2 in my code', Is correct? But the whole solving time is long.
ax=-200:0.01:200;
ay=2.^ax;
for i=1:SumsH
    f(i)=interp1(ax,ay,t_XiShu(i),'sos2');
end
Constraints=[Constraints, Log_Sum_ModeAP>=f'];

**Reply by johanlofberg** [reply]
f = 2^x is convex, hence f<= is convex so using an expensive sos2 model is unnecessary

and 2^200 just makes no sense so your range is way beyond reasonable

**Reply by BQKS** [reply]
But if not using sos2, gurobi can't solve the nonlinearities.

**Reply by johanlofberg** [reply]
based on what you have written here, 'graph' should work

**Reply by BQKS** [reply]
but using 'graph' ,i.e., f(i)=interp1(ax,ay,t_XiShu(i),'graph), f will contain the line of 2^x and the above of 2^x, then the solution of constraint Log_Sum_ModeAP>=f' will not consistence with original solution of Log_Sum_ModeAP>=2^x seriously.

**Reply by johanlofberg** [reply]
Unclear what you are saying

A sos2 model will create  a model f = max(a_i*x+b_i) representing the convex piecewise affine approximant. You want f <= Upper bound. A graph model will create a model which has f >=  max(a_i*x+b_i). You want f <= Upper bound, which thus is guaranteed to hold as max(a_i*x+b_i) <= f <= Upper bound.

**Comment by BQKS** [comment]
Yes, but if so, why not using 'sos2' directly to represent the approximation, which is closest to the convex than using 'graph' , because 'graph' will also produce the above of the convex function.

**Reply by johanlofberg** [reply]
Because sos2 is very expensive as it is a high-level mixed-integer representation. No reason to use a very expensive representation when a trivial linear model is sufficient

**Reply by BQKS** [reply]
of course, using 'graph '(f=interp1(ax,ay,t_XiShu(i),'graph')) will produce 2^x<=f<=Log_Sum_ModeAP, then if the solution satisfy f<=Log_Sum_ModeAP, the solution will also satisfy the original constraints Log_Sum_ModeAP>=2^x, but how much the upper limit of f, because f will contain +inf, then leading the unsolvable

**Reply by johanlofberg** [reply]
No, f can never become inf as you have constraints on f

Using a graph model will not allow any solutions which aren't feasible to the original problem, and allows all solutions in the original problem, i.e. the feasible set projected onto the original variables is the same. It is the whole idea of modelling in linear programming

For instance, you want to have the nonlinear constraint abs(x) <= 1. That can be written as f = max(x,-x) and f<=1 which you then model as the graph model x<= f, -x <= f, f<=1. Make sure you understand that the set of feasible x is exactly the same as in the original nonlinear constraint, despite the equality being replaced with inequality

**Reply by BQKS** [reply]
oh, I got it! Thank you very much!!!!!, but by the way the min function can use the 'graph ' to approximate?

**Reply by johanlofberg** [reply]
The function 'min' is concave and will thus be modelled using an hypograph when used in a convex fashion (i.e when bounding from below or maximizing) https://yalmip.github.io/command/min/

**Reply by BQKS** [reply]
But I have read a paper in the past that saying min is concave

**Reply by johanlofberg** [reply]
yes, that is exactly what I wrote

**Reply by BQKS** [reply]
oh, I made a mistake in writing, The paper said min is convex

**Reply by johanlofberg** [reply]
Well that is wrong. max is convex.

Don't confuse it with minimize and maximize though. Minimizing a convex function (such as max) is a convex problem. Maximizing a concave function (such as min) is a convex problem.

**Reply by BQKS** [reply]
oh, I got it, so the min{x,1-x} can view as convex, also can view as concave? Because x,1-x are Affine function？

**Reply by johanlofberg** [reply]
No,  min(1,1-x) is a concave function (draw it!), thus meaning the hypograph min(1,1-x) >= t is a convex set

**Reply by BQKS** [reply]
oh, it is concave (I draw it!), so should use interp1(ax,ay,t_XiShu(i),'hypograph’) to approximate min(1,1-x)?

**Reply by johanlofberg** [reply]
No, you only use 'graph'.  YALMIP will select suitable graph representation based on the function (and refuse if you don't follow rules required for a graph model to be possible)

However, if the function is min, you simply use that. interp1 is for general nonlinear functions which aren't already supported by yalmip

**Reply by BQKS** [reply]
Oh, but if using Gurobi, min should need use the "graph" to approximate

**Reply by johanlofberg** [reply]
If you use the min operator, regardless of solver used YALMIP will develop a graph model of that if possible, and if not possible it will create a MILP representation.

**Reply by BQKS** [reply]
Oh, so regardless of solver used, we can use min function directly, because Yamlap can tranform automaticly?

**Reply by johanlofberg** [reply]
Yes (with the caveat that if the solver doesn't support binaries, and the min operator is used in a nonconvex fashion, it will consequently not be possible)

**Reply by BQKS** [reply]
oh, I got it, thank you very much much muvh much!!!!Delayed your time so long!!

**Comment by BQKS** [comment]
Hi teacher, I use 'graph' to approximate exponential function in constraint, Gurobi show ''solve successfully'', but the solution obviously violates the approximational exponential constraint , whether the approximation funtion can not be identificated by Gurobi?

**Reply by johanlofberg** [reply]
you would have to supply a reproducible example

**Reply by BQKS** [reply]
oh, I got it, maybe because numerical problems in model rather than 'Interp1' using, because part of using 'Interp1' can satisfy the exponential constraint.

**Comment by BQKS** [comment]
Hi, teacher, it's also the question, when I use the 'graph' to approximate the exponential function, then Gurobi display ''Model contains large matrix coefficient range'', thus there are other way to solve the question? And the following is the my code.
ax=-15:1e-2:15;
ay=2.^ax;
for i=1:sumA
    f(i)=interp1(ax,ay,t_XiShu_ModeAP(i),'graph');
end
Constraints=[Constraints, Log_Sum_ModeAP>=f'];

**Reply by johanlofberg** [reply]
Yes, because the graph model of 2^x will basically be a straight line at x=15 with a very large derivative. You have a large span and will thus get huge numbers in the graph model

The develop branch of YALMIP now supports the nonlinear global stuff in gurobi 11, so using that you can use 2^x directly

**Reply by BQKS** [reply]
Ok, I try it, and solve successfully, thank you very much!

**Comment by BQKS** [comment]
Hello, teacher, I have a new question, constraint contains x^(1/2),  then the result dispalys: Solver not applicable (gurobi does not support signomial constraints),  whether it also because Yalmap doesn't suppot? Thanks!

**Comment by johanlofberg** [comment]
Yes that does not seem to have been added yet. You will have to write it using an auxilliary variable (assuming you target a nonlinear model and just haven't failed in setting vup asn socp-representable model)

**Reply by BQKS** [reply]
Thanks, I just add t>=x^(1/2)  to the constarint, then the problem can be solved!

**Comment by johanlofberg** [comment]
So how was it part of the model from the beginning if this suddenly works? This is still nonconvex and I thought I hadn't added support for that case yet

**Reply by BQKS** [reply]
The beginning of the model is sum()^2>=x, then I transform it to sum()>=x^(1/2), the result dispalys: Solver not applicable (gurobi does not support signomial constraints), then I add auxilliary variable  t>=x^(1/2) as you say, then the constraint become sum()>=t, t^2>=x, then it works.

**Reply by johanlofberg** [reply]
Aha, I thought you still used ^1/2. Yes that is the preferred way to model it anyway, as it leads to a scalar nonconvex QP constraint, instead of a general nonlinear constraint which still is is pretty immature in gurobi

**Reply by BQKS** [reply]
Yeah, actually, the original constraint sum()^2>=x can be solved directly, but because the data reason, the results display: Quadratic constrains contain large coefficient range, then I transform it to sum>=x^(1/2), the gurobi can't work, but actually using Taylor expansion of x^(1/2) to replace x^(1/2), then the constraint becomes LP, then can using iteration method to solve whole optimization problem, but in my model, x may take 0, thus the thought of using Taylor expansion can not work, then the constraint can must be solved by nonconvex, but thanks your before suggestions that using develop branch of YALMIP, it can solve the nonconvex!
